common/persistence/sql/sqlHistoryPersistence.go

```diff
@@ -0,0 +1,265 @@
+package sql
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/uber-common/bark"
+	workflow "github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common/persistence"
+
+	"github.com/hmgle/sqlx"
+)
+
+type (
+	sqlHistoryManager struct {
+		db      *sqlx.DB
+		shardID int
+		logger  bark.Logger
+	}
+
+	eventsRow struct {
+		DomainID     string  `db:"domain_id"`
+		WorkflowID   string  `db:"workflow_id"`
+		RunID        string  `db:"run_id"`
+		FirstEventID int64   `db:"first_event_id"`
+		Data         *[]byte `db:"data"`
+		DataEncoding string  `db:"data_encoding"`
+		DataVersion  int64   `db:"data_version"`
+
+		RangeID int64 `db:"range_id"`
+		TxID    int64 `db:"tx_id"`
+	}
+)
+
+const (
+	appendHistorySQLQueryTemplate = `INSERT IGNORE INTO events (
```

Comment:
How do you insert a row and conditionally fail if the row already exists?

A plain insert will cause an error, but we cannot distinguish that error from other errors without inspecting the error message

INSERT IGNORE will "succeed" if the row already exists but not modify any rows. Hence, on success, we can check the number of rows affected to determine whether the INSERT actually happened.

The alternative was to SELECT then INSERT, but we would require some kind of "gap lock" in between